### PR TITLE
Reverse geolocation webhook node 

### PR DESCRIPTION
**Pull request by aharshal623**
## Summary
 Target issue is #3616 
This PR is the first attempt at taking on the PRD for reverse geocoding. In this PR, we use the Google Maps API to return city, state, country, etc. for a given lat/long. Additionally, we implemented functionality for the API key to be taken from .env.dev and not be publicly exposed.

### CODE REVIEW

File: lib/glific/clients/common_webhook.ex [hunk 1]
```diff
@@ -311,6 +358,14 @@ defmodule Glific.Clients.CommonWebhook do
     end
   end
 
+  defp find_component(components, type) do
```
Comment (madclaws) [inline lib/glific/clients/common_webhook.ex:337]: Can we add function spec here?

File: lib/glific/clients/common_webhook.ex [hunk 2]
```diff
@@ -302,6 +302,53 @@ defmodule Glific.Clients.CommonWebhook do
   def webhook("check_response", fields),
     do: %{response: String.equivalent?(fields["correct_response"], fields["user_response"])}
 
+  def webhook("geolocation", fields) do
```
Comment (madclaws) [inline lib/glific/clients/common_webhook.ex:279]: Can we add test case for this function?

**Review by madclaws — CHANGES_REQUESTED** [review]

**Comment by madclaws** [comment]
Can you also fix the credo warnings @aharshal623

File: lib/glific/communications/message.ex [hunk 3]
```diff
@@ -191,8 +191,6 @@ defmodule Glific.Communications.Message do
   """
   @spec receive_message(map(), atom()) :: :ok | {:error, String.t()}
   def receive_message(%{organization_id: organization_id} = message_params, type \\ :text) do
-    Logger.info("Received message: type: '#{type}', id: '#{message_params[:bsp_message_id]}'")
```
Comment (AkhileshNegi) [inline lib/glific/communications/message.ex]: can you revert these changes and take the latest pull

File: lib/glific/communications/wa_group_message.ex [hunk 4]
```diff
@@ -80,8 +80,6 @@ defmodule Glific.Communications.GroupMessage do
   """
   @spec receive_message(map(), atom()) :: :ok | {:error, String.t()}
   def receive_message(%{organization_id: organization_id} = message_params, type \\ :text) do
-    Logger.info("Received message: type: '#{type}', id: '#{message_params[:bsp_id]}'")
```
Comment (AkhileshNegi) [inline lib/glific/communications/wa_group_message.ex]: can you revert these changes and take the latest pull

File: test/glific/flows/webhook_test.exs [hunk 5]
```diff
@@ -248,4 +259,63 @@ defmodule Glific.Flows.WebhookTest do
...
+    result = CommonWebhook.webhook("geolocation", fields)
+
+    assert result[:success] != nil
+    assert result[:city] != nil
+    assert result[:state] != nil
+    assert result[:country] != nil
+    assert result[:postal_code] != nil
+    assert result[:district] != nil
+    assert result[:ward] != nil
+    assert result[:address] != nil
```
Comment (AkhileshNegi) [inline test/glific/flows/webhook_test.exs]: can u match the actual values rather than matching not nil

**Review by AkhileshNegi — CHANGES_REQUESTED** [review]

File: test/glific/flows/webhook_test.exs [hunk 6]
```diff
@@ -248,4 +259,63 @@ defmodule Glific.Flows.WebhookTest do
...
+        %Tesla.Env{
+          status: 500,
+          body: "Internal Server Error"
+        }
+    end)
+
+    result = CommonWebhook.webhook("geolocation", fields)
+
+    # Assert that success is false and an error message is returned
+    assert result[:success] == false
```
Comment (madclaws) [inline test/glific/flows/webhook_test.exs]: you can use `refute result[:success]` instead of this line.

File: test/glific/flows/webhook_test.exs [hunk 7]
```diff
@@ -248,4 +259,63 @@ defmodule Glific.Flows.WebhookTest do
...
+    assert result[:error] != nil
```
Comment (madclaws) [inline test/glific/flows/webhook_test.exs]: can use `is_nil`

File: lib/glific/clients/common_webhook.ex [hunk 8]
```diff
@@ -311,6 +359,15 @@ defmodule Glific.Clients.CommonWebhook do
     end
   end
 
+  @spec find_component(list(map()), String.t()) :: String.t()
+  defp find_component(components, type) do
+    components
+    |> Enum.find(fn component ->
+      type in component["types"]
+    end)
+    |> then(&((&1 && &1["long_name"]) || "N/A"))
```
Comment (madclaws) [inline lib/glific/clients/common_webhook.ex]: Can we make this simple, it looks way cryptic for reading. @aharshal623

**Review by madclaws — CHANGES_REQUESTED** [review]

**Review by madclaws — APPROVED** [review]

**Pull request merged**